plugins/flashrom/fu-flashrom-device.c

```diff
@@ -48,11 +72,206 @@ fu_flashrom_device_set_quirk_kv (FuDevice *device,
+	gint rc;
+
+	if (flashrom_init (SELFCHECK_TRUE)) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "flashrom initialization error");
+		return FALSE;
+	}
+	flashrom_set_log_callback (fu_flashrom_device_debug_cb);
+	if (flashrom_programmer_init (&priv->flashprog, "internal", NULL)) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "programmer initialization failed");
+		return FALSE;
+	}
+	rc = flashrom_flash_probe (&priv->flashctx, priv->flashprog, NULL);
+	if (rc == 3) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "flash probe failed: multiple chips were found");
+		return FALSE;
+	}
+	if (rc == 2) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "flash probe failed: no chip was found");
+		return FALSE;
+	}
+	if (rc != 0) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "flash probe failed: unknown error");
+		return FALSE;
+	}
+	priv->flash_size = flashrom_flash_getsize (priv->flashctx);
+	if (priv->flash_size == 0) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "flash size zero");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_flashrom_device_prepare (FuDevice *device,
+			  FwupdInstallFlags flags,
+			  GError **error)
+{
+	FuFlashromDevicePrivate *priv = GET_PRIVATE (FU_FLASHROM_DEVICE (device));
+	g_autofree gchar *firmware_orig = NULL;
+	g_autofree gchar *basename = NULL;
+
+	/* if the original firmware doesn't exist, grab it now */
+	basename = g_strdup_printf ("flashrom-%s.bin", fu_device_get_id (device));
+	firmware_orig = g_build_filename (FWUPD_LOCALSTATEDIR, "lib", "fwupd",
+					  "builder", basename, NULL);
+	if (!fu_common_mkdir_parent (firmware_orig, error))
+		return FALSE;
+	if (!g_file_test (firmware_orig, G_FILE_TEST_EXISTS)) {
+		g_autofree guint8 *newcontents = g_malloc0 (priv->flash_size);
+		g_autoptr(GBytes) buf = NULL;
+
+		fu_device_set_status (device, FWUPD_STATUS_DEVICE_READ);
+		if (flashrom_image_read (priv->flashctx, newcontents, priv->flash_size)) {
+			g_set_error_literal (error,
+					     FWUPD_ERROR,
+					     FWUPD_ERROR_READ,
+					     "failed to back up original firmware");
+			return FALSE;
+		}
+		buf = g_bytes_new_static (newcontents, priv->flash_size);
+		if (!fu_common_set_contents_bytes (firmware_orig, buf, error))
+			return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_flashrom_device_write_firmware (FuDevice *device,
+				   FuFirmware *firmware,
+				   FwupdInstallFlags flags,
+				   GError **error)
+{
+	FuFlashromDevicePrivate *priv = GET_PRIVATE (FU_FLASHROM_DEVICE (device));
+	gsize sz = 0;
+	gint rc;
+	const guint8 *buf;
+	g_autoptr(GBytes) blob_fw;
+
+	blob_fw = fu_firmware_get_bytes (firmware, error);
```

Comment:
combine these lines to fix the CI failure.

Reply:
Thanks! Done